src/commands/giveaway.rs

```diff
@@ -0,0 +1,650 @@
+use chrono::{DateTime, Utc};
+use rand::seq::SliceRandom;
+use std::collections::HashMap;
+
+use serenity::{
+    builder::{CreateComponents, CreateEmbed, CreateEmbedFooter},
+    model::{
+        interactions::{
+            message_component::ButtonStyle, InteractionApplicationCommandCallbackDataFlags,
+            InteractionResponseType,
+        },
+        prelude::application_command::*,
+    },
+    prelude::*,
+};
+
+use crate::{
+    database::Database, extensions::*, models::Giveaway, Http, Interaction, Message, ReactionType,
+    User,
+};
+
+static DEFAULT_DURATION: i64 = 3600;
+static DEFAULT_WINNERS: i64 = 1;
+static DEFAULT_PRIZE: &str = "Nothing :(";
+pub static GIVEAWAY_REACTION_CHAR: char = '🎉';
+
+struct ListOffset;
+
+impl TypeMapKey for ListOffset {
+    type Value = HashMap<u64, i64>;
+}
+
+async fn ensure_offset_map(ctx: &Context) {
+    let mut data = ctx.data.write().await;
+    if !data.contains_key::<ListOffset>() {
+        data.insert::<ListOffset>(HashMap::new());
+    }
+}
+
+fn footer_with_text<S: Into<String>>(text: S) -> CreateEmbedFooter {
+    let mut footer = CreateEmbedFooter {
+        0: std::collections::HashMap::new(),
+    };
+    footer.text(text.into());
+    footer
+}
```

Comment:
I don't really see the point of this helper, It's quite easy to just do:
```rs
embed.footer(|f| f.text(text.into()))
```

Also could all of these some `Create****`-helpers just be one function, like in /src/voting.rs
if some of them are always used together.